### PR TITLE
Make use of queues and add blinking platforms 

### PR DESCRIPTION
**Pull request by Jontpan**
When testing, use basicPlatformScheme(30, 40, 5, 15, 10);

#96

### CODE REVIEW

File: 2Dgame/Assets/Scripts/Platform Scripts/BlinkerScript.cs [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+        Rigidbody2D rb = collision.rigidbody;
+        float playerHalf = rb.transform.localScale.x / 2;
+        float dist = (rb.position.y - playerHalf) - this.transform.position.y;
+
+        if (rb.velocity.y <= 0.1f && dist >= -0.12f)
+        {
+            rb.velocity = new Vector2(rb.velocity.x, jumpSpeed);
+            if (GameControllerScript.soundOn)
+            {
+                audioSource.Play();
```
Comment (AxlLind) [inline 2Dgame/Assets/Scripts/Platform Scripts/BlinkerScript.cs]: This appears to give a null pointer exception when you jump on the platform

File: 2Dgame/Assets/Scripts/Platform Scripts/BlinkerScript.cs [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+    new Renderer renderer;
+
+    private new void Awake()
+    {
+        renderer = GetComponent<Renderer>();
+    }
+
+    /**
+     * Enables the renderer when jumped on
```
Comment (AxlLind) [inline 2Dgame/Assets/Scripts/Platform Scripts/BlinkerScript.cs]: This doesn't seem to happen? Is the idea for them to be shown again when you jump on them again? When I've played they stay hidden.

File: 2Dgame/Assets/Scripts/PlatformSpawnerScript.cs [hunk 3]
```diff
@@ -40,83 +45,81 @@ public class PlatformSpawnerScript : MonoBehaviour {
...
      * Difficulty increases as score increases.
      */
-    GameObject choosePlatform()
+    void choosePlatform()
     {
         if (this.transform.position.y < 200)
         {
-            return basicPlatformScheme();
+            basicPlatformScheme();
+            return;
```
Comment (AxlLind) [inline 2Dgame/Assets/Scripts/PlatformSpawnerScript.cs]: These return statements are now redundant

**Review by AxlLind — COMMENTED** [review]
Looks good! Some small things I've left comments on but there is one major thing:

When there are two blinker platforms on the screen when you jump on the first one the next blinker's platforms get hidden as well.

Also if you've already jumped on a blinker (purple one) and then an other appears the hidden ones start blinking when jumping on the new blinker.

File: 2Dgame/Assets/Scripts/Platform Scripts/BlinkerScript.cs (hunk 2 above)
Comment (Jontpan) [reply 2Dgame/Assets/Scripts/Platform Scripts/BlinkerScript.cs]: Check to see if the blinkerprefab has the correct script attached, I thought i fixed it in the add correct script commit. I may have unfixed it when i reset the basicplatformscheme....

File: 2Dgame/Assets/Scripts/PlatformSpawnerScript.cs (hunk 3 above)
Comment (Jontpan) [reply 2Dgame/Assets/Scripts/PlatformSpawnerScript.cs]: You are correct. For a minute there I thought the bug I had was that it fell through the if statements.

File: 2Dgame/Assets/Scripts/Platform Scripts/BlinkerScript.cs (hunk 2 above)
Comment (Jontpan) [reply 2Dgame/Assets/Scripts/Platform Scripts/BlinkerScript.cs]: Don't seem to have this problem, you probably don't have the latest commits on this branch

File: 2Dgame/Assets/Scripts/Platform Scripts/BlinkerScript.cs (hunk 1 above)
Comment (Jontpan) [reply 2Dgame/Assets/Scripts/Platform Scripts/BlinkerScript.cs]: Also something I believe I fixed in some of the later commits.

**Comment by Jontpan** [comment]
Yep, I outlined these problems in the commit message. While I have some idea of how to fix it, I hoped that it would never occur... :P 

*sigh*

I'll get on it.

**Comment by AxlLind** [comment]
My bad, didn't read the commit message. Found a problem with the prefabqueue implementation, though. It keeps adding stuff to the queue even when there are stuff in the queue. This would be ok if we only added one prefab and took one out every time but with the flip and blinker platform we add more than one prefab and only take out one.

This means the queue increases in length quickly. It always increases and never decreases. With some quick testing it was up to 50 elements at y = 200.

**Comment by Jontpan** [comment]
I am aware of this but figured it wouldn't be too bad. But we could honestly just clear the queue at every difficulty increase.

**Comment by AxlLind** [comment]
Cant we just add an if statement around the choosePlatform()?

``` C#
if (prefabQueue.Count == 0) 
{
    choosePlatform();
}
```

That way we don't add stuff to the queue unless we need to.

**Comment by Jontpan** [comment]
Lot better.

File: 2Dgame/Assets/Scripts/Platform Scripts/OnJumpScript.cs [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+using System.Collections.Generic;
+using UnityEngine;
+
+public abstract class OnJumpScript : PlatformScript {
+
+    protected bool notJumpedOn;
+
+    protected new void Awake()
+    {
+        notJumpedOn = false;
```
Comment (AxlLind) [inline 2Dgame/Assets/Scripts/Platform Scripts/OnJumpScript.cs]: This needs to be set to true.

File: 2Dgame/Assets/Scripts/PlatformSpawnerScript.cs [hunk 5]
```diff
@@ -40,83 +49,75 @@ public class PlatformSpawnerScript : MonoBehaviour {
...
         }
         else if (this.transform.position.y < 1000)
         {
-            return PlatformScheme(60, 20, 10, 10); // Flips
+            PlatformScheme(60, 20, 10, 10); // Flips
         }
         else if (this.transform.position.y < 1500)
         {
-            return PlatformScheme(45, 40, 5, 10); // More hard stuff
+            PlatformScheme(30, 40, 5, 15, 10); // More hard stuff, first blink
```
Comment (AxlLind) [inline 2Dgame/Assets/Scripts/PlatformSpawnerScript.cs:116]: I feel like blinkers could happen earlier maybe

File: 2Dgame/Assets/Scripts/Platform Scripts/BlinkingPlatformScript.cs [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+        }
+    }
+
+    /**
+     * Blinks away the next three blinkers
+     */
+    protected override void OnJump()
+    {
+        GameObject[] blinkers = GameObject.FindGameObjectsWithTag("Blinker");
+        StartCoroutine(gcs.PauseForSeconds(2));
```
Comment (AxlLind) [inline 2Dgame/Assets/Scripts/Platform Scripts/BlinkingPlatformScript.cs]: Maybe this should be slightly shorter? I feel like it's fine for the flips but for the blinkers I think it should maybe be shorter.

File: 2Dgame/2Dgame.sln [hunk 7]
```diff
@@ -1,29 +1,26 @@
...
-		{F93AA5A7-4DB8-E983-2867-D1607B9B74DF}.Release|Any CPU.Build.0 = Release|Any CPU
-		{34C318DC-2A81-229A-44B8-AB4598E102A1}.Debug|Any CPU.ActiveCfg = Debug|Any CPU
-		{34C318DC-2A81-229A-44B8-AB4598E102A1}.Debug|Any CPU.Build.0 = Debug|Any CPU
-		{34C318DC-2A81-229A-44B8-AB4598E102A1}.Release|Any CPU.ActiveCfg = Release|Any CPU
-		{34C318DC-2A81-229A-44B8-AB4598E102A1}.Release|Any CPU.Build.0 = Release|Any CPU
+		{3A334A53-003A-5D3C-1187-9E32BAE4F6C3}.Debug|Any CPU.ActiveCfg = Debug|Any CPU
+		{3A334A53-003A-5D3C-1187-9E32BAE4F6C3}.Debug|Any CPU.Build.0 = Debug|Any CPU
+		{3A334A53-003A-5D3C-1187-9E32BAE4F6C3}.Release|Any CPU.ActiveCfg = Release|Any CPU
+		{3A334A53-003A-5D3C-1187-9E32BAE4F6C3}.Release|Any CPU.Build.0 = Release|Any CPU
+		{DE6F5ADE-BF29-988E-0D90-29C150F7701D}.Debug|Any CPU.ActiveCfg = Debug|Any CPU
```
Comment (AxlLind) [inline 2Dgame/2Dgame.sln:18]: Shouldn't this be "DE6F5ADE-BF29-988E-0D90-29C150F7701G"?

File: 2Dgame/Assets/Scripts/Scene Controllers/GameControllerScript.cs [hunk 8]
```diff
@@ -80,9 +81,18 @@ public class GameControllerScript : MonoBehaviour {
 		playerRB.simulated = isPaused ? false : true;
 	}
 
-	// Button press-methods
+    public IEnumerator PauseForSeconds(float seconds)
```
Comment (AxlLind) [inline 2Dgame/Assets/Scripts/Scene Controllers/GameControllerScript.cs:84]: This is a really nice function!

**Review by AxlLind — APPROVED** [review]
Looks really good! I really like the blinking effect 👍  The code looks much better now with the queue and with the OnJump() change. Nice work!

Just fix the NotJumpedOn to true instead of false and we're good!

File: 2Dgame/Assets/Scripts/PlatformSpawnerScript.cs (hunk 5 above)
Comment (Jontpan) [reply 2Dgame/Assets/Scripts/PlatformSpawnerScript.cs:116]: we should probably decide together exactly how the progression should be, maybe when we finalized the game for release. I'll leave it right now.

File: 2Dgame/2Dgame.sln (hunk 7 above)
Comment (Jontpan) [reply 2Dgame/2Dgame.sln:18]: o shit haha. how long has this bug been here lol

File: 2Dgame/Assets/Scripts/Platform Scripts/BlinkingPlatformScript.cs (hunk 6 above)
Comment (Jontpan) [reply 2Dgame/Assets/Scripts/Platform Scripts/BlinkingPlatformScript.cs]: Tested, 1.2 seconds feels good.

**Pull request merged**